### PR TITLE
[SPARK-32432][SQL] Add support for reading ORC/Parquet files of SymlinkTextInputFormat table And Fix Analyze for SymlinkTextInputFormat table

### PR DESCRIPTION
**Pull request by CHENXCHEN**
### What changes were proposed in this pull request?

1. Add support for reading ORC/Parquet files of SymlinkTextInputFormat table
2. Fix analyze table size for SymlinkTextInputFormat

### Why are the changes needed?

1. Both trino(prestosql) and prestodb support reading ORC/Parquet files of SymlinkTextInputFormat table, but Spark does not, Spark support other formats except ORC/Parquet format, and this PR add support for reading ORC/Parquet files of SymlinkTextInputFormat table
2. As follows example, analyze the table size will be 100 instead of 19999, which affects the join optimization, and it will attempt to deliver a large table broadcast if `manifest` file size is less than `spark.sql.autoBroadcastJoinThreshold`, and after this PR, analyze the table size will be 19999

we have files:
```
size   filepath
100    hdfs:///path/to/table/manifest
9999   hdfs:///path/to/other/part-1.parquet.orc
10000  hdfs:///path/to/other/part-2.parquet.orc
```

content of `hdfs:///path/to/table/manifest` :
```
hdfs:///path/to/other/part-1.parquet.orc
hdfs:///path/to/other/part-2.parquet.orc
```

table ddl:
```
CREATE EXTERNAL TABLE symlink_orc ( name STRING, version DOUBLE, sort INT )
ROW FORMAT SERDE 'org.apache.hadoop.hive.ql.io.orc.OrcSerde'
STORED AS INPUTFORMAT 'org.apache.hadoop.hive.ql.io.SymlinkTextInputFormat'
OUTPUTFORMAT 'org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat'
LOCATION 'hdfs:///path/to/table';
```

See details in the JIRA. [SPARK-32432](https://issues.apache.org/jira/browse/SPARK-32432)

### Does this PR introduce _any_ user-facing change?
Yes
1. Before this pr, an exception was thrown when reading the ORC/Parquet files of SymlinkTextInputFormat table.
2. Analyze SymlinkTextInputFormat table, will parse its real file size(such as ORC/Parquet files size), not the `manifest` file size, as described above 

### How was this patch tested?

Added Unit Test: org.apache.spark.sql.hive.SymlinkSuite

### CODE REVIEW

**Comment by CHENXCHEN** [comment]
Could you please take a look and review this PR ?

**Comment by CHENXCHEN** [comment]
This PR is a bit like [this](https://github.com/apache/spark/pull/29330) but is too old to support Spark 3.2.
In addition, this PR add a fix to the analysis table size.

**Comment by CHENXCHEN** [comment]
cc @cloud-fan @gengliangwang @viirya @sadikovi

**Comment by CHENXCHEN** [comment]
ok to test

**Comment by CHENXCHEN** [comment]
cc @cloud-fan  could you help take a look when you have time? Thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala [hunk 1]
```diff
@@ -74,21 +74,44 @@ object CommandUtils extends Logging {
...
+      calculateMultipleLocationSizes(spark, catalogTable.identifier,
+        getPartitionPaths(partitions, isSymlinkTable, fs)).sum
     }
     logInfo(s"It took ${(System.nanoTime() - startTime) / (1000 * 1000)} ms to calculate" +
       s" the total size for table ${catalogTable.identifier}.")
     totalSize
   }
 
+  def getPartitionPaths(partitions: Seq[CatalogTablePartition]
+                        , isSymlinkTable: Boolean, fs: => FileSystem): Seq[Option[URI]] = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala:103]:
```suggestion
  def getPartitionPaths(
      partitions: Seq[CatalogTablePartition],
      isSymlinkTable: Boolean,
      fs: => FileSystem): Seq[Option[URI]] = {
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala [hunk 2]
```diff
@@ -74,21 +74,44 @@ object CommandUtils extends Logging {
...
+    partitions.flatMap { catalogTablePartition =>
+        if (isSymlinkTable) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala:105]: nit: 2 spaces indentation

**Comment by cloud-fan** [comment]
I think the problem today is we don't have a good abstraction for this feature at the framework level. This is a special hive file format that changes the behavior of file listing, while in Spark the `FileIndex` API assumes the file listing behavior is unrelated to the file format.

This PR simply adds special handling of `SymlinkTextInputFormat` in several places, and I'm OK with it if it's super hard to come up with a good abstraction for it, but we should give it a try first.

cc @viirya @dongjoon-hyun @AngersZhuuuu

**Comment by viirya** [comment]
Same feeling here. Not sure if `SymlinkTextInputFormat` is the only case with custom file listing behavior.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala [hunk 3]
```diff
@@ -122,7 +122,16 @@ abstract class PartitioningAwareFileIndex(
               // Directory does not exist, or has no children files
               Nil
           }
-          PartitionDirectory(values, files)
+          // Check leaf files since they might be symlink targets
+          if (files.isEmpty) {
+            val status: Seq[FileStatus] = leafFiles.get(path) match {
```
[the remark this answers is not in the document]
Comment (CHENXCHEN) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala:127]: `SymlinkTextInputFormat` table specify the list of files for a table/partition based on the content of a text file.
As the example in the PR description shows, the real data file is not in the table root

[the remark this answers is not in the document]
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala:127]: > the real data file is not in the table root

Yea, but `leafFiles` are listed from table root, how symlink targets might be in leaf files?

[the remark this answers is not in the document]
Comment (CHENXCHEN) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala:127]: At the partition table, we have the same behavior, there is also symbolic in the root directory of the partition

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 4]
```diff
@@ -204,10 +204,23 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
         (options, None)
       }
 
+    lazy val fs = tablePath.getFileSystem(sparkSession.sparkContext.hadoopConfiguration)
+    val isSymlinkTextFormat = SymlinkTextInputFormatUtil.isSymlinkTextFormat(relation.tableMeta)
+
+    val symlinkTargets = if (isSymlinkTextFormat) {
+      SymlinkTextInputFormatUtil.getTargetPathsFromSymlink(fs, tablePath)
```
Comment (AngersZhuuuu) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala:211]: When table is partitioned and `lazyPruningEnabled ` = false, this value is not used, I think we don't need to get this value earliear

**Comment by AngersZhuuuu** [comment]
Agree with @cloud-fan that we don't have a good abstraction for this feature.
IMO, we can  add a new method under `CatalogTable` and `CatalogTablePartition`, such as
```
getFileList(fileFormat: String， fs: FileSystem)
```

Then we can directly list files according to file format, when in all other place, we can just call this API.  Then we don't need to judge if it's  `SymlinkTextFormat` every where.

WDYT @cloud-fan @viirya @dongjoon-hyun

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 5]
```diff
@@ -220,6 +233,8 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
 
         if (paths.isEmpty) {
           Seq(tablePath)
+        } else if (isSymlinkTextFormat) {
```
Comment (CHENXCHEN) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala:236]: @AngersZhuuuu also used in here when `lazyPruningEnabled` = false

**Pull request closed without merging**